cpp/include/raft/neighbors/brute_force.cuh

```diff
@@ -138,13 +138,15 @@ inline void knn_merge_parts(
  * 					 is ignored if the metric_type is not Minkowski.
  * @param[in] global_id_offset: optional starting global id mapping for the local partition
  *                              (assumes the index contains contiguous ids in the global id space)
+ * @param[in] distance_epilogue: optional epilogue function to run after computing distances
  */
 template <typename idx_t,
           typename value_t,
           typename value_int,
           typename matrix_idx,
           typename index_layout,
-          typename search_layout>
+          typename search_layout,
+          typename epilogue_op = raft::identity_op>
```

Comment:
I guess in the current implementation (at least as far as I can tell from the diff), we are not fusing the epilogue to the pairwise distances but instead calling it as a transform after the fact. I think this will probably be the best for the hdbscan case, anyways, since there's some random access lookups that slow down the pw dist computation otherwise. 

Besides, if we did fuse the epilogue to the pw dists, hdbscan would propagate quite far down into the distances, which would mean the precompiled specializations would no longer going to be used. 


Tagging @ahendriksen @teju85 @tfeher for their thoughts here. This seems like the right approach to me- it means launching another kernel but the needed logic for hdbscan's reachability computation is not a trivial mapping.

Reply:
> which would mean the precompiled specializations would no longer going to be used.
Indeed.

I need a bit more context to weigh in. The epilogue is K-selection? Or building the MST?

Reply:
The epilogue for hdbscan is this code https://github.com/rapidsai/cuml/blob/bed5213e3a671f18dcfdf2bae912508a17bd24da/cpp/src/hdbscan/detail/reachability.cuh#L167-L176  

We're not fusing it to the pairwise distances, or the select_k op right now (though we could in the future) -